feat(babel): add process.env.EXPO_OS

EvanBacon opened the pull request:
# Why

When working with lower-level tooling, you don't want to be importing `react-native` everywhere. This PR enables you to access the bundler setting `platform` via `process.env.EXPO_OS`, which is also inlined (improved tree-shaking).

The docs for the Expo Babel preset also seem to have relevant versioning info which makes them a good candidate for moving to the official docs. 

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

# How

- Inline `process.env.EXPO_OS` with the platform that is being bundled.
- Move docs from readme into the versioned config docs https://docs.expo.dev/versions/latest/config/babel/

<!--
How did you build this feature or fix this bug and why?
-->

# Test Plan

- Added tests in the babel plugin.

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

EvanBacon committed:
feat(babel): add support for `process.env.EXPO_OS`

EvanBacon committed:
version docs

EvanBacon committed:
Update babel.mdx

EvanBacon committed:
Apply suggestions from code review

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

alanjhughes committed:
[sharing][ios] Migrate to Expo Modules (#20969)

Simek commented on docs/pages/versions/unversioned/config/babel.mdx:
```diff
@@ -0,0 +1,162 @@
+
+| Value     | Description             |
+| --------- | ----------------------- |
+| `ios`     | Runs on iOS devices     |
+| `android` | Runs on Android devices |
```
```suggestion
| `android` | Runs on Android devices |
| `ios`     | Runs on iOS devices     |
```

Platforms order nit.

Simek commented on docs/pages/versions/unversioned/config/babel.mdx:
```diff
@@ -0,0 +1,162 @@
+
+## Babel Caller
+
+> This section is for advanced developers who want to use `babel-preset-expo` with a custom bundler implementation.
```
```suggestion
> **info** This section is for advanced developers who want to use `babel-preset-expo` with a custom bundler implementation.
```

Let's emphasise note a bit.

Simek commented on packages/babel-preset-expo/README.md:
````diff
@@ -1,129 +1,3 @@
-    }
-],
-```
+> The docs have moved to [Expo: `babel.config.js`](https://docs.expo.dev/versions/latest/config/babel/).
````
```suggestion
> **Note**
> The documentation has been moved to [`babel.config.js` page](https://docs.expo.dev/versions/latest/config/babel/) in Expo Docs.
```

GFM formatting nit + rephrase.

Simek commented on docs/pages/versions/unversioned/config/babel.mdx:
```diff
@@ -0,0 +1,162 @@
+
+Enabling this option will allow your project to run with older JavaScript syntax (i.e. `module.exports`). This option will break tree shaking and increase your bundle size, but will eliminate the following error when `module.exports` is used:
+
+> `TypeError: Cannot assign to read only property 'exports' of object '#<Object>'`
```
Maybe the raw code block (```) will fit better there than note?

Simek commented on docs/pages/versions/unversioned/config/babel.mdx:
```diff
@@ -0,0 +1,162 @@
+
+Changes Babel's compiled `import` statements to be lazily evaluated when their imported bindings are used for the first time.
+
+_Note:_ this option has an effect only when the `disableImportExportTransform` option is set to `false`. On Android and iOS, `disableImportExportTransform` defaults to `false`, and on web it defaults to `true` to allow for tree shaking.
```
```suggestion
> This option has an effect only when the `disableImportExportTransform` option is set to `false`. On Android and iOS, `disableImportExportTransform` defaults to `false`, and on web it defaults to `true` to allow for tree shaking.
```

Let's use note component in there instead of italic sentence start.

Simek commented on docs/pages/versions/unversioned/config/babel.mdx:
```diff
@@ -0,0 +1,162 @@
+
+This property is passed down to [`@babel/plugin-transform-react-jsx`](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx). This options does nothing when `jsxRuntime` is not set to `automatic`.
+
+### [`lazyImports`](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy)
```
Not sue it this link will work or if it will, if it will be correct. In general it's recommended to not use links inside of the headings.

Simek commented on docs/pages/versions/unversioned/config/babel.mdx:
```diff
@@ -0,0 +1,162 @@
+
+The value of `lazyImports` has a few possible effects:
+
+- `null` - [metro-react-native-babel-preset](https://github.com/facebook/metro/tree/master/packages/metro-react-native-babel-preset) will handle it. (Learn more about it here: https://github.com/facebook/metro/commit/23e3503dde5f914f3e642ef214f508d0a699851d)
+- `false` - No lazy initialization of any imported module.
+- `true` - Lazy-init all imported modules except local imports (e.g., `./foo`), certain Expo packages that have side effects, and the two cases mentioned [here](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy).
+- `string[]` - [babel-plugin-transform-modules-commonjs](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy) will handle it.
+- `(string) => boolean` - [babel-plugin-transform-modules-commonjs](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy) will handle it.
+  If you choose to do this, you can also access the list of Expo packages that have side effects by using `const lazyImportsBlacklist = require('babel-preset-expo/lazy-imports-blacklist');` which returns a `Set`.
```
Probably table will be a bit more readable than list.

Simek commented on docs/pages/versions/unversioned/config/babel.mdx:
```diff
@@ -0,0 +1,162 @@
+
+Projects using `babel-preset-expo`, have access to the following environment variables:
+
+- `process.env.EXPO_OS`: The operating system of the device. This can be used as a replacement for `Platform.OS`. One of `ios`, `android`, `web`.
```
```suggestion
- `process.env.EXPO_OS`: The operating system of the device. This can be used as a replacement for `Platform.OS`. One of `android`, `ios`, `web`.
```

Platform order nit.

Simek reviewed (commented):
Few comments and suggestion on the text content part of the PR.

gabrieldonadel committed:
[docs] Update Video snack example to work with projects using TS (#21925)

# Why

While triaging some issues on our repo, I noticed [this
report](https://github.com/expo/expo/issues/21913) regarding an error
when using the `resizeMode` prop with `expo-av`'s `Video` component. The
error reported is expected and only happens in projects using
Typescript, as the `resizeMode` prop expects a value from the
`ResizeMode` enum, but our docs examples are using a string value
directly, which can be misleading for users. By updating the example to
use the `ResizeMode` enum instead of the string value, we can prevent
this issue from occurring and improve the user experience.


# How

Update the `Video` snack example to use the `ResizeMode` enum instead of
using the string value directly

# Test Plan

N / A

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg:
updated a module plugin).

Kudo committed:
[packages][templates] update react-native 0.71.6 (#21909)

update react-native 0.71.6 for sdk 48. the changelog is at https://github.com/facebook/react-native/releases/tag/v0.71.5 and https://github.com/facebook/react-native/releases/tag/v0.71.6 which has some meaningful fixes.

update package versions

ci passed

Kudo committed:
[build-properties][templates] add enableShrinkResourcesInReleaseBuilds support (#21911)

# Why

add `enableShrinkResourcesInReleaseBuilds` build property to support `shrinkResources` android build setting
close ENG-8017

# How

add `enableShrinkResourcesInReleaseBuilds` build property and integrate to template

# Test Plan

- unit tests passed
- test on a managed project

Kudo committed:
et ssbc -b sdk-48 [skip ci]

Kudo committed:
Publish expo-build-properties

expo-build-properties@0.5.2

EvanBacon committed:
Apply suggestions from code review

Co-authored-by: Bartosz Kaszubowski <gosimek@gmail.com>

EvanBacon committed:
update jest to provide platform

wschurman commented on packages/expo-updates/CHANGELOG.md:
```diff
@@ -40,6 +38,13 @@
+
+### 🐛 Bug fixes
+
+- Change arg in gradle `.execute()` call to null to inherit env variables from user's env ([#21712](https://github.com/expo/expo/pull/21712) by [@phoenixiguess](https://github.com/phoenixiguess))
```
unrelated to this PR? how did we create a release with just these changes (manual release cherry-picking?) cc @douglowder

tsapeta reviewed (commented):
It seems to be messed up with other commits 😅

EvanBacon commented:
Closing as we can't predictably ensure platforms are defined in Jest at the moment. Will reproach in the future.
